pep_sphinx_extensions/pep_processor/transforms/pep_contents.py

```diff
@@ -45,19 +44,34 @@ def __init__(self, document, startnode=None):
         self.backlinks = None
 
     def apply(self) -> None:
-        # used in parts.Contents.build_contents
-        self.toc_id = self.startnode.parent["ids"][0]
-        self.backlinks = self.document.settings.toc_backlinks
-
-        # let the writer (or output software) build the contents list?
-        if getattr(self.document.settings, "use_latex_toc", False):
-            # move customisation settings to the parent node
-            self.startnode.parent.attributes.update(self.startnode.details)
-            self.startnode.parent.remove(self.startnode)
+        contents = self.build_contents(self.document[0][4:])  # skip PEP title, headers, <hr/>, and contents
+        if contents:
+            self.startnode.replace_self(contents)
         else:
-            contents = self.build_contents(self.document[0])
-            if contents:
-                self.startnode.replace_self(contents)
-            else:
-                # if no contents, remove the empty placeholder
-                self.startnode.parent.parent.remove(self.startnode.parent)
+            # if no contents, remove the empty placeholder
+            self.startnode.parent.parent.remove(self.startnode.parent)
+
+    def build_contents(self, node, level=0):
+        level += 1
+        entries = []
+        children = getattr(node, "children", node)
+
+        for section in children:
+            if not isinstance(section, nodes.section):
+                continue
+
+            title = section[0]
+            ref_id = section['ids'][0]
+            if title.next_node(nodes.reference) is None:
+                title["refid"] = ref_id  # Add a link to self
+            entry_text = self.copy_and_filter(title)
+            reference = nodes.reference("", "", refid=ref_id, *entry_text)
+            item = nodes.list_item("", nodes.paragraph("", "", reference))
+
+            if level < 1_000_000:
```

Comment:
Why 1 million? Can you capture this into a constant?

Reply:
In the docutils source it was sys.maxsize, and that was a bit too high - in reality I don't know how much this function recurses but 1 million seemed a nice "big number" that wasn't too big

Reply:
The alternate is to remove the check and let the internal recursion limit handle excessive recursion?

Reply:
The less code the better, but if you want to leave it, add it as a constant at the global level with some name and add some comment :)

Reply:
I'm not attached - will remove

Reply:
done